Chatkit/Auth/TokenProvider.swift

```diff
@@ -0,0 +1,139 @@
+public class HTTPSTokenProvider: TokenProvider {
+    
+    private let method: String
+    private let host: String
+    private let path: String
+    private let getHeaders: AsyncCredentialsCall?
+    private let getQueryParams: AsyncCredentialsCall?
+    
+    /// Create an HTTPSTokenProvider which presents either headers or query parameters as part of
+    /// the request. These should be used to identify your application user session to your backend
+    /// so that it can issue a token for the user.
+    ///
+    /// - Parameters:
+    ///     - method: The HTTP method to use in the call, e.g. `POST`
+    ///     - host: The host name to be called, e.g. `example.com`
+    ///     - path: The path component of the URL to be called, e.g. `/tokens`
+    ///     - headers: An optional map of headers to include in the request. Here you can supply
+    ///     the session or other credientials which your endpoint might require to authenticate
+    ///     the request.
+    ///     - queryParams: An optional map of query parameters to include in the request URL.
+    ///     Here you can supply the session or other credientials which your endpoint might
+    ///     require to authenticate the request.
+    public convenience init(method: String,
+                            host: String,
+                            path: String,
+                            headers: [String: String]?,
+                            queryParams: ([String: String]?)) {
+        self.init(method: method,
+                  host: host,
+                  path: path,
+                  getHeaders: HTTPSTokenProvider.wrapCredentialsValue(headers),
+                  getQueryParams: HTTPSTokenProvider.wrapCredentialsValue(queryParams))
+    }
+    
+    /// Create an HTTPSTokenProvider which presents either headers or query parameters as part of
+    /// the request. These should be used to identify your application user session to your backend
+    /// so that it can issue a token for the user.
+    ///
+    /// - Parameters:
+    ///     - method: The HTTP method to use in the call, e.g. `POST`
+    ///     - host: The host name to be called, e.g. `example.com`
+    ///     - path: The path component of the URL to be called, e.g. `/tokens`
+    ///     - headers: An optional async function which will be invoked when a request is about
+    ///     to be made, so that you can supply headers describing the current application user
+    ///     session.
+    ///     - queryParams: An optional async function which will be invoked when a request is about
+    ///     to be made, so that you can supply query parameters describing the current application
+    ///     user session.
+    public init(method: String,
+                host: String,
+                path: String,
+                getHeaders: AsyncCredentialsCall?,
+                getQueryParams: AsyncCredentialsCall?) {
```

Comment:
Wouldn't it be worth to call it more generic terms, e.g. "KeyValue" instead of "Credentials"? After all it could be that there's an auth header that can expire or get invalidated but query params still are needed and have nothing to do with auth (probably they're static too but that's another issue with trying to aim for such generic implementation).

Still not sure how we can fit in notifying about the need to refresh of the auth token. Still leaning [towards](https://github.com/pusher/chatkit-swift/pull/228#issuecomment-563986272) the most.